libraries/ESP8266mDNS/src/LEAmDNS2Host_Transfer.cpp

```diff
@@ -0,0 +1,2470 @@
+    return bResult;
+}
+#endif
+
+#ifdef MDNS_IPV6_SUPPORT
+/*
+    MDNSResponder::_buildDomainForReverseIPv6
+
+    The IPv6 address is stringized by printing the 16 address bytes (32 nibbles) into a char buffer in reverse order
+    and adding 'ip6.arpa' (eg. 3.B.6.E.A.1.B.B.A.B.F.7.F.8.0.1.0.0.0.0.0.0.0.0.0.0.0.0.0.8.E.F.ip6.arpa).
+    Used while detecting reverse IPv6 questions and answering these
+
+*/
+bool clsLEAMDNSHost::_buildDomainForReverseIPv6(IPAddress p_IPv6Address,
+        clsLEAMDNSHost::clsRRDomain& p_rReverseIPv6Domain) const
+{
+    bool    bResult = ((p_IPv6Address.isSet()) &&
+                       (p_IPv6Address.isV6()));
+
+    p_rReverseIPv6Domain.clear();
+
+    const uint16_t* pRaw = p_IPv6Address.raw6();
+    for (int8_t i8 = (clsConsts::u16IPv6Size / 2); ((bResult) && (i8 > 0)); --i8) // 8..1
+    {
+        uint16_t	u16Part = ntohs(pRaw[i8 - 1] & 0xFFFF);
+        char        acBuffer[2];
+        for (uint8_t u8 = 0; ((bResult) && (u8 < 4)); ++u8)             // 0..3
+        {
+            itoa((u16Part & 0xF), acBuffer, 16);
+            bResult = p_rReverseIPv6Domain.addLabel(acBuffer);
+            u16Part >>= 4;
+        }
+    }
+    bResult = ((bResult) &&
+               (p_rReverseIPv6Domain.addLabel(clsConsts::pcReverseIPv6Domain)) &&    // .ip6.arpa
+               (p_rReverseIPv6Domain.addLabel(clsConsts::pcReverseTopDomain)) &&     // .local
+               (p_rReverseIPv6Domain.addLabel(0)));
+    DEBUG_EX_ERR(if (!bResult) DEBUG_OUTPUT.printf_P(PSTR("%s _buildDomainForReverseIPv6: FAILED!\n"), _DH()););
+    return bResult;
+}
+#endif
+
+
+/*
+
+    UDP
+
+*/
+
+/*
+    MDNSResponder::_udpReadBuffer
+
+*/
+bool clsLEAMDNSHost::_udpReadBuffer(unsigned char* p_pBuffer,
+                                    size_t p_stLength)
+{
+    bool    bResult = ((m_pUDPContext->getSize() >= p_stLength) &&
+                       (p_pBuffer) &&
+                       (p_stLength) &&
+                       ((p_stLength == m_pUDPContext->read((char*)p_pBuffer, p_stLength))));
+    DEBUG_EX_ERR(if (!bResult) DEBUG_OUTPUT.printf_P(PSTR("%s _udpReadBuffer: FAILED!\n"), _DH()););
+    return bResult;
+}
+
+/*
+    MDNSResponder::_udpRead8
+
+*/
+bool clsLEAMDNSHost::_udpRead8(uint8_t& p_ru8Value)
+{
+    return _udpReadBuffer((unsigned char*)&p_ru8Value, sizeof(p_ru8Value));
+}
+
+/*
+    MDNSResponder::_udpRead16
+
+*/
+bool clsLEAMDNSHost::_udpRead16(uint16_t& p_ru16Value)
+{
+    bool    bResult = false;
+
+    if (_udpReadBuffer((unsigned char*)&p_ru16Value, sizeof(p_ru16Value)))
+    {
+        p_ru16Value = lwip_ntohs(p_ru16Value);
+        bResult = true;
+    }
+    return bResult;
+}
+
+/*
+    MDNSResponder::_udpRead32
+
+*/
+bool clsLEAMDNSHost::_udpRead32(uint32_t& p_ru32Value)
+{
+    bool    bResult = false;
+
+    if (_udpReadBuffer((unsigned char*)&p_ru32Value, sizeof(p_ru32Value)))
+    {
+        p_ru32Value = lwip_ntohl(p_ru32Value);
+        bResult = true;
+    }
+    return bResult;
+}
```

Comment:
I was thinking along the following:
functions overloaded to map hton and ntoh:
```cpp
uint8_t  ntoh(uint8_t  value) {return value;} //a noop
uint16_t ntoh(uint16_t value) {return lwip_ntohs(value);}
uint32_t ntoh(uint32_t value) {return lwip_ntohl(value);}

uint8_t  hton(uint8_t value) {return value;}  //a noop
uint16_t hton(uint16_t value) {return lwip_htons(value);}
uint32_t hton(uint32_t value) {return lwip_htonl(value);}
```

Then template _udpRead:
```cpp
template<typename T>
bool clsLEAMDNSHost::_udpRead(T &value)
{
    if(_udpReadBuffer(unsiged char *)&value, sizeof(value))
    {
        value = ntoh(value);
        return true;
    }
    return false;
}
```

Similar templates for _udpAppend() and _write(). You replace 9 functions with 6 one-liners and 2 templates.

HOWEVER: further down there are calls to write16() and write8() etc (see _writeMDNSMsgHeader() ). I think in this case it may be better to _not_ template the methods, so that the write()s are explicit, i.e.: you can see at a glance what type is being written. Having that at a glance allows the reader to think about the protocol data sequence.

Reply:
Isn't that a better practice ?
```
uint8_t  ntoh(uint8_t  value)
{
    return value;
} //a noop

uint16_t ntoh(uint16_t value)
{
    return lwip_ntohs(value);
}

uint32_t ntoh(uint32_t value)
{
    return lwip_ntohl(value);
}

uint8_t  hton(uint8_t value)
{
    return value;
}  //a noop

uint16_t hton(uint16_t value)
{
    return lwip_htons(value);
}

uint32_t hton(uint32_t value)
{
    return lwip_htonl(value);
}
```

Joke apart, **humans** need to have ntoh/hton functions with the byte number in their name.
These functions deal with low level libraries, structures or protocols, not with upper level c++ classy classes where everything fit with magic.